File: cli/src/transformerlab_cli/util/chunked_upload.py
"""Reusable chunked-upload-of-one-file helper used by task, model, and dataset commands.

Uses the generic /upload/{init,chunk,status,complete} endpoints in the API. Returns the
upload_id of the assembled file so callers can plumb it through to a domain-specific
endpoint (e.g. /model/fileupload?upload_id=...).
"""

import math
import os
from typing import Optional

from rich.progress import Progress, TaskID

import transformerlab_cli.util.api as api


def upload_one_file(
    local_path: str,
    *,
    server_filename: Optional[str] = None,
    progress: Optional[Progress] = None,
    progress_task: Optional[TaskID] = None,
) -> str:
    """Upload one file via /upload/init → /chunk → /complete and return upload_id.

    Resume-aware: if /upload/{id}/status reports already-received chunks, those
    chunk indices are skipped.

    `server_filename`, if provided, is what the server records in the staged
    metadata (used for filename rather than relpath; relpath is passed
    separately to the domain endpoint).

    `progress` and `progress_task` are optional: when set, the helper advances
    the rich progress task by one per uploaded chunk.
    """
    size = os.path.getsize(local_path)
    filename = server_filename or os.path.basename(local_path)

    init_resp = api.post_json(
        "/upload/init",
        json_data={"filename": filename, "total_size": size},
    )
    if init_resp.status_code != 200:
        raise RuntimeError(f"upload init failed ({init_resp.status_code}): {init_resp.text}")
    init_body = init_resp.json()
    upload_id = init_body["upload_id"]
    chunk_size = init_body["chunk_size"]
    total_chunks = math.ceil(size / chunk_size)

    status_resp = api.get(f"/upload/{upload_id}/status")
    received: set[int] = set(status_resp.json().get("received", [])) if status_resp.status_code == 200 else set()
    if progress is not None and progress_task is not None:
        progress.advance(progress_task, len(received))

    with open(local_path, "rb") as fh:
        for i in range(total_chunks):
            if i in received:
                continue
            fh.seek(i * chunk_size)
            chunk = fh.read(chunk_size)
            put_resp = api.put(
                f"/upload/{upload_id}/chunk?chunk_index={i}",
                content=chunk,
                headers={"Content-Type": "application/octet-stream"},
            )
            if put_resp.status_code != 200:
                raise RuntimeError(f"chunk {i} failed ({put_resp.status_code}): {put_resp.text}")
            if progress is not None and progress_task is not None:
                progress.advance(progress_task, 1)

    complete_resp = api.post_json(
        f"/upload/{upload_id}/complete",
        json_data={"total_chunks": total_chunks},
        timeout=None,
    )
    if complete_resp.status_code != 200:
        raise RuntimeError(f"upload complete failed ({complete_resp.status_code}): {complete_resp.text}")
    return upload_id

File: src/renderer/lib/chunkedUpload.ts
import * as chatAPI from './transformerlab-api-sdk';

const CHUNK_SIZE = 64 * 1024 * 1024; // 64 MB

export interface ChunkedUploadOptions {
  file: Blob;
  filename: string;
  onProgress?: (percent: number) => void;
}

export interface ChunkedUploadResult {
  upload_id: string;
  temp_path: string;
}

/**
 * Upload a file in 64MB chunks via the /upload API.
 * Returns the upload_id to pass to the consuming endpoint as ?upload_id=.
 * Throws on any failure.
 */
export async function chunkedUpload({
  file,
  filename,
  onProgress,
}: ChunkedUploadOptions): Promise<ChunkedUploadResult> {
  const totalSize = file.size;
  const totalChunks = Math.ceil(totalSize / CHUNK_SIZE);

  // 1. Init
  const initResp = await chatAPI.authenticatedFetch(
    chatAPI.Endpoints.Upload.Init(),
    {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ filename, total_size: totalSize }),
    },
  );
  if (!initResp.ok) throw new Error(`Upload init failed: ${initResp.status}`);
  const { upload_id } = await initResp.json();

  // 2. Check for already-received chunks (resumability)
  const statusResp = await chatAPI.authenticatedFetch(
    chatAPI.Endpoints.Upload.Status(upload_id),
  );
  const statusBody = statusResp.ok ? await statusResp.json() : { received: [] };
  const alreadyReceived: Set<number> = new Set(statusBody.received ?? []);

  // 3. Send chunks
  for (let i = 0; i < totalChunks; i++) {
    if (alreadyReceived.has(i)) {
      onProgress?.(Math.round(((i + 1) / totalChunks) * 100));
      continue;
    }
    const start = i * CHUNK_SIZE;
    const end = Math.min(start + CHUNK_SIZE, totalSize);
    const chunk = file.slice(start, end);
    const url = chatAPI.Endpoints.Upload.Chunk(upload_id) + `?chunk_index=${i}`;
    const chunkResp = await chatAPI.authenticatedFetch(url, {
      method: 'PUT',
      headers: { 'Content-Type': 'application/octet-stream' },
      body: chunk,
    });
    if (!chunkResp.ok)
      throw new Error(`Chunk ${i} upload failed: ${chunkResp.status}`);
    onProgress?.(Math.round(((i + 1) / totalChunks) * 100));
  }

  // 4. Complete
  const completeResp = await chatAPI.authenticatedFetch(
    chatAPI.Endpoints.Upload.Complete(upload_id),
    {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ total_chunks: totalChunks }),
    },
  );
  if (!completeResp.ok)
    throw new Error(`Upload complete failed: ${completeResp.status}`);
  const { temp_path } = await completeResp.json();

  return { upload_id, temp_path };
}

export async function deleteUpload(upload_id: string): Promise<void> {
  await chatAPI.authenticatedFetch(chatAPI.Endpoints.Upload.Delete(upload_id), {
    method: 'DELETE',
  });
}
